### PR TITLE
cache adherence data

### PR DESCRIPTION
**Pull request by pbugni**
Significant change to adherence reporting mechanism.
Generating a report is now very fast, but wholly dependent on cached data, as what is cached is used without taking the expensive time to determine if it's still valid.

There are scheduled jobs to maintain the cache, set to run daily.  Necessary to build the initial cache, and thereafter, to pick up anything that falls through the cracks.

Any event that alters a user's timeline (submission of QNR, organization altering RP, change of consent ...) will also invalidate the user's adherence report, and trigger a regeneration on next request for the timeline.

The /timeline view for a patient will now also display the user's cached adherence data.  `?purge=all` will force a recalculation of everything, including adherence data.

### CODE REVIEW

**Review by mcjustin — APPROVED** [review]

File: tests/test_reporting.py [hunk 1]
```diff
@@ -132,6 +134,60 @@ def test_protected_results(self):
...
+            },
+            "Month 12": {
+                "qb": "CRV Baseline v2",
+                "site": "CRV",
+                "visit": "Month 12",
+                "status": "Overdue",
+                "consent": "20 - May - 2023 07: 42:46 ",
+                "country ": None,
+                "user_id ": 3,
+                "study_id": "study user 3",
```
Comment (ivan-c) [inline tests/test_reporting.py:161]: Are these normally numeric, and this is just fake data?

**Review by ivan-c — APPROVED** [review]
Looks good! Had a few questions, but ready to merge

File: tests/test_reporting.py (hunk 1 above)
Comment (pbugni) [reply tests/test_reporting.py:161]: just fake test data - typically the `study_id` is ###-##.

File: portal/models/adherence_data.py [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+
+
+class AdherenceData(db.Model):
+    """ Cached adherence report data
+
+    Full history adherence data is expensive to generate, retain between reports.
+    Cache reportable data in simple JSON structure, maintaining keys for lookup
+    and invalidation timestamps.
+
+    rs_id_visit: the numeric rs_id and visit month string joined with a colon
```
[the remark this answers is not in the document]
Comment (pbugni) [reply portal/models/adherence_data.py:16]: they have little value independently, are unique and valuable as a set.  we could use a `composite key` on the DB to manage, but this format made all the queries easier to work with.

File: portal/factories/app.py [hunk 3]
```diff
@@ -137,6 +137,11 @@ def configure_app(app, config):
     """Load successive configs - overriding defaults"""
     app.config.from_object(DefaultConfig)
     app.config.from_pyfile('base.cfg', silent=True)
+    site_cfg_fullpath = os.path.join(app.config.root_path, SITE_CFG)
+    if not os.path.isfile(site_cfg_fullpath):
+        app.logger.warn(f"{site_cfg_fullpath} not found!!!!!!!!!!!!!")
+    else:
+        app.logger.warn(f"{site_cfg_fullpath} FOUND!!!!!!!!!!!!!!")
```
[the remark this answers is not in the document]
Comment (pbugni) [reply portal/factories/app.py:144]: agreed - debugging cruft, thanks for noticing.

**Review by ivan-c — APPROVED** [review]
Looks good, thanks! Will merge once tests pass, unless you beat me to it

**Comment by pbugni** [comment]
@ivan-c hold up!  `logger.warn` is deprecated in place of `logger.warning` - will fix.

**Pull request merged**